2단계 - 연관 관계 매핑 리뷰 요청드립니다.

KwakTaeMin opened the pull request:
연관 관계 매핑 해보았습니다. 리뷰 잘 부탁드립니다~

Commit:
docs : 연관 관계 매핑 (2단계 요구사항)

Commit:
feat : Answer FK 매핑 (User, Question)
- FK Join Column 추가 (writer_id, question_id)
- QnAService 위 사항으로 코드 수정
- AnswerRepositoryTest test code 수정

Commit:
feat : Question FK 매핑 (User)
- FK Join Column 추가 (writer_id)
- QnAService 위 사항으로 코드 수정
- QuestionRepositoryTest test code 수정

Commit:
feat : DeleteHistory FK 매핑 (User)
- FK Join Column 추가 (deleted_by_id)
- QnAService 위 사항으로 코드 수정
- DeleteHistoryRepositoryTest test code 수정

JunilHwang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -4,27 +4,32 @@
 	@Column(name = "deleted", nullable = false)
 	private boolean deleted = false;
 
+	@ManyToOne
+	@JoinColumn(name = "question_id", foreignKey = @ForeignKey(name = "fk_answer_to_question"))
+	private Question question;
+
+	@ManyToOne
+	@JoinColumn(name = "writer_id", foreignKey = @ForeignKey(name = "fk_answer_writer"))
+	private User writer;
```
lazyLoad 관련 설정도 추가해주면 어떨까요?

KwakTaeMin replied:
지연로딩(Lazy Load), 즉시로딩(Eager Load)에 대해서 공부할 수 있어서 좋았고,
즉시로딩은 지양하는 것과 @ManyToXXX 는 default가 즉시로딩(Eager Load)여서 수정하라는 의미로 받아드려도 될까요?

그리고 질문있는데요! 프록시 객체에서 제 테스트 코드 중

```
@Test
	@DisplayName("Answer id로 Answer 조회 테스트")
	void findById() {
		Answer answer = new Answer(writer, question, "내용");
		Answer saved = answers.save(answer);
		Answer actual = answers.findByIdAndDeletedFalse(saved.getId()).get();
		assertAll(
			() -> assertThat(actual).isNotNull(),
			() -> assertThat(actual.getContents()).isEqualTo(answer.getContents()),
			() -> assertThat(actual.isDeleted()).isEqualTo(false),
			() -> assertThat(actual.getCreateAt()).isNotNull(),
			() -> assertThat(actual.getWriter()).isEqualTo(answer.getWriter()),
			() -> assertThat(actual.getWriter().getName()).isEqualTo("name"),
			() -> assertThat(actual.getWriter().getPassword()).isEqualTo("password"),
			() -> assertThat(actual.getWriter().getUserId()).isEqualTo("javajigi"),
			() -> assertThat(actual.getWriter().getEmail()).isEqualTo("javajigi@slipp.net"),
			() -> assertThat(actual.getQuestion().getContents()).isEqualTo("contents1"),
			() -> assertThat(actual.getQuestion().getTitle()).isEqualTo("title1")
		);
	}
```

지연로드로 했을 경우에, User 테이블이나, Question 테이블에 왜 쿼리가 가지 않는지,
혹시 가지않는다면, User나 Question이 save를 한 후에 1차 캐시라는 영역에 있고, DB에 쿼리를 날린 후,
변경되지 않아 1차 캐시에 그대로 있는 상태이며 
조회했을 때 프록시 객체에서도 1차 캐시 영역에서 가져와서 진행되는 것인지 궁금합니다.
제가 알고 있는 이러한 개념이 맞는지 아니면 엉뚱한 생각을 하는건지 궁금합니다!

JunilHwang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -4,27 +4,32 @@
 	private boolean deleted = false;
 
+	@ManyToOne
+	@JoinColumn(name = "question_id", foreignKey = @ForeignKey(name = "fk_answer_to_question"))
+	private Question question;
```
`@Where` Annotation을 사용해보세요!

KwakTaeMin replied:
`@Where` 절을 사용했더니 deleted가 false인걸 default로 가져올 수 있네요! 좋은 것 같아요 감사합니다!

JunilHwang reviewed (changes requested):
안녕하세요 태민님! 연관관계 매핑 미션 잘 진행해주셨네요 👍
간단한 리뷰 몇 가지 남겼으니 확인 후 다시 피드백 요청 부탁드립니다 🙇‍♂️

KwakTaeMin committed:
refactor : Answer Entity 리팩토링
- Question LazyLoad 적용, @Where 어노테이션을 사용하여 deleted flase를 기본으로 조회
- User : LazyLoad 적용

KwakTaeMin committed:
refactor : Answer Entity 리팩토링
- Question LazyLoad 적용, @Where 어노테이션을 사용하여 deleted false를 기본으로 조회
- User : LazyLoad 적용

KwakTaeMin committed:
style : Answer Entity Space Trim

Commit:
refactor : DeleteHistory, Question Lazy Load 수정

Commit:
refactor : Question @Where 절 추가

Commit:
refactor : Question, Answer
- ColumnDefault deleted false로 설정

JunilHwang reviewed (approved):
안녕하세요 태민님! 피드백 반영 잘 해주셨네요 👍 

> 지연로딩(Lazy Load), 즉시로딩(Eager Load)에 대해서 공부할 수 있어서 좋았고, 즉시로딩은 지양하는 것과 @ManyToXXX 는 default가 즉시로딩(Eager Load)여서 수정하라는 의미로 받아드려도 될까요?

즉시로딩을 하게 될 경우, 불필요한 쿼리 비용이 생길 수 있기 때문입니다. 지금은 2~3개의 테이블이 엮여있는데 만약에 5개, 10개 정도의 테이블이 엮여있고 즉시로딩을 한다고 치면 비용이 무척 크지 않을까요? 꼭 필요한 경우가 아니라면 지연로딩을 사용하는게 좋습니다!

> 지연로드로 했을 경우에, User 테이블이나, Question 테이블에 왜 쿼리가 가지 않는지

결론부터 말씀 드리면, 쿼리는 발생합니다. 다만 `getWriter()`를 호출하는 시점에 select 쿼리가 발생한답니다. 그리고 한 번의 select가 발생한 후에 데이터의 변화가 없는 상태에서 다시 `getWriter()`를 할 경우에 캐시에서 가져오게 된답니다.

조금 더 자세한 내용은 https://ict-nroo.tistory.com/132 여기를 참고해주세요!

Step2는 여기서 마무리하겠습니다 😁
